### ISSUE TITLE
Controller was executed twice when navigated through ngLink

### ISSUE THREAD

**Issue by shaunxu** [angular/router#204]
I found when I clicked a link generated by ngLink, the targeting controller was being invoked twice. This is not happened when I visit that page directly (type URL or refresh the page), just when clicking the ngLink.

I just created a small sample to reproduce this issue at https://github.com/shaunxu/ng-new-router-test/ Just clone, `npm install` and then run `node --port [the port number you like] --debug`

Don't know if it's because of something I did wrong.

**Comment by skusunam** [comment]
Is this duplicate of #200?

**Comment by shaunxu** [comment]
@skusunam Seems not. In my case I only have one ng-viewport in my `index.html`.

**Comment by skusunam** [comment]
@shaunxu thanks for clarifying. If you can provide a plunker example (or) running gh-pages it will help to run it and see the problem right away rather than cloning your repo and going through whole 9 yards.

**Comment by shaunxu** [comment]
@skusunam Sure. Just noticed plunker support folders so I will copy my example.

**Comment by shaunxu** [comment]
@skusunam Just created a plunker http://plnkr.co/edit/FAageDG6OlU12Vs26DN2?p=preview Please open browser console, it shows that controller was executed twice when clicked a link.

**Comment by skusunam** [comment]
@shahata I can clearly see this problem with your pluker and also from "examples/angular-1/hello/" this. @btford once you confirm this as "bug" may be i can dig in and see if i can do a PR on this. Any hints will help me in doing this first code PR :)

**Comment by rohanrehman** [comment]
@skusunam have you found a workaround?

**Comment by robertdp** [comment]
It seems to be an issue in `routerFactory()`.

https://github.com/angular/router/blob/master/src/router-directive.es5.js#L85

`$$rootRouter.navigate()` updates `$location.path()`, which triggers the `$rootScope.$watch`, which in turn calls `$$rootRouter.navigate()` again...

It doesn't just instantiate the controllers twice, it runs through the whole route activation pipeline twice.

**Comment by rohanrehman** [comment]
@btford @robertdp that was what I found too

**Comment by basvandenheuvel** [comment]
Is there a solution yet?

**Comment by Jcamilorada** [comment]
is there any workaround?

**Comment by tomer78** [comment]
I found a solution. 
I don't use the ng-link.
I've  used an ng-click and I redirect using the $location service. 
 this.$location.path('new url')

This way the controller instantiation  happens only once!

**Comment by wsfriday-sfd** [comment]
My solution was similar to @tomer78 , but I've modified ngLinkDirective and anchorLinkDirective directly to use $location.path() to change routes instead of using $router.navigate() and triggering this double controller issue

```javascript
		function ngLinkDirective($router, $location, $parse) {
					 // we can avoid adding a watcher if it's a literal
					 if (routeParamsGetter.constant) {
						 var params = routeParamsGetter();
		-        url = '.' + router.generate(routeName, params);
		+        url = router.generate(routeName, params);
						 elt.attr('href', url);
					 } else {
						 scope.$watch(function() {
							 return routeParamsGetter(scope);
						 }, function(params) {
		-          url = '.' + router.generate(routeName, params);
		+          url = router.generate(routeName, params);
							 elt.attr('href', url);
						 }, true);
					 }
				 } else {
		-      url = '.' + router.generate(routeName);
		+      url = router.generate(routeName);
					 elt.attr('href', url);
				 }
		}

		function ngLinkDirective($router, $location, $parse) {
		 ngLinkDirective.$inject = ["$router", "$location", "$parse"];
 
 
		-function anchorLinkDirective($router) {
		+function anchorLinkDirective($router, $location) {
			 return {
				 restrict: 'E',
				 link: function(scope, element) {
						function anchorLinkDirective($router) {
							 event.preventDefault();
						 }
						 if ($router.recognize(href)) {
		-          $router.navigate(href);
		+          $location.path(href);
							 event.preventDefault();
						 }
					 });
				 }
			 }
		 }
		-anchorLinkDirective.$inject = ["$router"];
		+anchorLinkDirective.$inject = ["$router", "$location"];
```

**Comment by dteunkenstt** [comment]
The solution @wsfriday-sfd proposed didn't quite do it for me.

The problem I observed was that the url first was e.g. './home' and then '/home' (note the missing dot in the second url)

I modified router.es5.js somewhere else. I didn't modify any lines, I just added a few.

```javascript
(createClass)(Router, {
  ...
  navigate: function(url) {
      var $__0 = this;
      if (this.navigating) {
        return $q.when();
      }
      this.lastNavigationAttempt = url;
      var instruction = this.recognize(url);
      if (!instruction) {
        return $q.reject();
      }
+     if(this.lastInstructionCanonicalUrl === instruction.canonicalUrl) {
+       return $q.reject();  
+     }
+     this.lastInstructionCanonicalUrl = instruction.canonicalUrl;
      this._startNavigating();
      ...
```

The `instruction.canonicalUrl` is the path, without the '.' in front (the one @wsfriday-sfd seems to be removing).

**Comment by smartworld-dm** [comment]
<link href="/assets/css/style.css" rel="stylesheet" type="text/css">
<div ng-init="activeUser()">
</div>

I've got same issue, When using link the activeUser() is called twice, but when I use the direct address, it is called once.
